E2E tests setup with laravel dusk

vsergiu93 opened the pull request:
This PR does the following:

1. Sets up Laravel Dusk with Sail
2. Add E2E tests for home page
3. Add E2E tests for login 
4. Add E2E tests for register
5. Add some guidelines on how to get started with this setup in README.md
6. Sets up github action to run Laravel Dusk tests


Closes https://github.com/brendt/rfc-vote/issues/231

brendt commented:
Wow, this looks great! Anything you want my input on?

vsergiu93 commented:
@brendt Thanks. Yes feel free to suggest improvements to what I have so far.
Right now there is at least one thing that is problematic with this PR, that is making it work on apple silicon, here is the link to the official sail documentation about [Laravel Dusk](https://laravel.com/docs/10.x/sail#laravel-dusk), as we can see there are two distinct images that need to be used based on what cpu type one has, ideally we would like to make this as seamless as possible, maybe we can make the docker image dynamic with some env variable. I haven't tried the solution.

I am open to solutions on how to solve the selenium docker image issue, not sure if this PR should be blocked by this issue.

What I would like to do in this PR is the following: 

- [ ] setup laravel dusk with sail (partially done, see the issue described above)
- [X] setup github workflow for laravel dusk ( this is done, ready for some feedback)
- [ ] add e2e tests for homepage 
- [ ] add e2e tests for login
- [ ] add e2e tests for register
- [ ] add some documentation about running laravel dusk with sail

joshbonnick commented:
Can we add an `@dusk` directive instead of adding `id` attributes to the HTML just for testing?

```php
Blade::directive('dusk', function ($expression) {
    if ( ! app()->isProduction()) {
        return "<?php echo 'dusk=\"' . $expression . '\"'; ?>";
    }

    return null;
});
```

Then in the blade files `<div class="some-classes" @dusk('main-header')></div>`. The directive will add the `dusk` attribute to elements but not pollute the HTML in production.

vsergiu93 commented:
@joshbonnick Honestly I'm ok with any suggestion, the way I started to add tests was to use this [page objects](https://laravel.com/docs/10.x/dusk#generating-pages) which also allows you to map some css selectors to some dusk specific selector.

Honestly I'm not sure what would be the best way to handle this. But I can say I didn't think of adding a custom blade directive.

vsergiu93 commented:
> Can we add an `@dusk` directive instead of adding `id` attributes to the HTML just for testing?
> 
> ```
> Blade::directive('dusk', function ($expression) {
>     if ( ! app()->isProduction()) {
>         return "<?php echo 'dusk=\"' . $expression . '\"'; ?>";
>     }
> 
>     return null;
> });
> ```
> 
> Then in the blade files `<div class="some-classes" @dusk('main-header')></div>`. The directive will add the `dusk` attribute to elements but not pollute the HTML in production.

@joshbonnick I tried to add this, but I don't think this is valid in blade, I mean you can't use the directive in the way you suggested, maybe I'm just not familiar with this. In any case I not sure if we should bother that much with removing the dusk selector in prod.

vsergiu93 commented:
> In the dusk env file
> 
> ```
> SELENIUM_IMAGE=selenium/standalone-chromium
> ```
Thanks, I added this, it is working, but just one thing, it needs to be added to .env file, I think sail looks at .env when you run the commands like `sail up -d`

> You should also add `BCRYPT_ROUNDS` to the Dusk env file. Which will speed up tests which hash passwords, [further explanation here](https://laravel-news.com/tips-to-speed-up-phpunit-tests) in 'Password hash rounds'
> 
> ```
> BCRYPT_ROUNDS=4
> ```
I added this, indeed I saw some small improvements in test time, thanks for suggesting it.

vsergiu93 commented on .phpstorm.meta.php:
```diff
@@ -0,0 +1,7 @@
+
+namespace PHPSTORM_META {
+
+    override(\Laravel\Dusk\Browser::visit(0), type(0));
```
This is a weird trick to improve a bit the developer experience with [page methods](https://laravel.com/docs/10.x/dusk#page-methods).

Basically with this PhpStorm doesn't complain about undefined method, when doing `visit(new SomePage)` followed by some methods define in SomePage, here is an example from LoginPageTest: 
![image](https://github.com/brendt/rfc-vote/assets/5923843/89d659af-3ca0-4fb0-b592-b0fc310ed139)

@brendt please let me know how you feel about this, it is a bit unconventional and I wasn't able to find a better way a more general way of improving the page methods.

brendt replied:
I'm totally fine with it. I use the Laravel Idea plugin, which I think supports this out of the box, but I'm fine adding it via phpstorm meta as well 👍

vsergiu93 replied:
I use Laravel Idea plugin as well, but I wasn't able to make it work :)

vsergiu93 commented:
@brendt This is ready for review, please have a look at my previous comments as well, also not really sure what happened with github history, we can try to fix it, I somehow messed up when I tried to sync the fork .

Thanks

Erol314 commented:
> not really sure what happened with github history, we can try to fix it, I somehow messed up when I tried to sync the fork .
> 

try rebasing your branch from main locally, and push force the branch

vsergiu93 commented:
> > not really sure what happened with github history, we can try to fix it, I somehow messed up when I tried to sync the fork .
> 
> try rebasing your branch from main locally, and push force the branch

I think it is better now thanks, need to be a bit more careful in the future with fork syncing

vsergiu93 commented:
@brendt Do I need to do anything special to get Qodana to pass? :)

brendt commented on resources/views/layouts/base.blade.php:
```diff
@@ -32,12 +32,12 @@ class="scroll-smooth"
         x-data="{ open: false }"
     >
-        <div class="dark:text-gray-200 text-lg md:text-xl font-bold relative z-20">
+        <div dusk="header-logo" class="dark:text-gray-200 text-lg md:text-xl font-bold relative z-20">
```
Shouldn't we use `@dusk('header-logo)`. I kind of thought Laravel had this built-in, but apparently it hasn't? 🤔

The idea would be that `@dusk` will only write the `x-dusk` attribute when not in production, so that we don't have additional dusk attributes on the real site.

vsergiu93 replied:
Hmm, I'll try to look more into this, I haven't found anything here https://laravel.com/docs/10.x/dusk#dusk-selectors

brendt commented:
> I tried to add this, but I don't think this is valid in blade, I mean you can't use the directive in the way you suggested, maybe I'm just not familiar with this. In any case I not sure if we should bother that much with removing the dusk selector in prod.

Maybe we should just do it via a function?

```php
function dusk(string $name): ?HtmlString
{
    if (app()->isProduction()) {
        return null;
    }
    
    return new HtmlString($name);
}
```

```blade
<div {{ dusk('header-logo') }} 
```

brendt commented:
Apart from the dusk attribute, this looks all super nice! I'm happy to merge it when the attribute thing has been fixed.

vsergiu93 commented on app/Helpers/Dusk.php:
```diff
@@ -0,0 +1,14 @@
+use Illuminate\Support\HtmlString;
+
+if (! function_exists('dusk')) {
+    function dusk(string $selector): ?HtmlString
```
@brendt I tried with this helper function, it works on html elements but not on blade components :(. 
This is the error I get when I attempt to use this helper on a blade component. 
```balde
        <x-home.title {{dusk('test')}}>
            {{ __('Open RFCs') }}
        </x-home.title>
```
![image](https://github.com/brendt/rfc-vote/assets/5923843/3d1ddb8e-5252-460e-ac92-55a6dbf64f71)

Not really sure how else to approach this at this point :), I'm willing to work more on this but I would need some ideas/suggestions.

brendt replied:
Well I think the dusk function should be move to the component itself, no? Otherwise you'd have to repeat it every time you add the component.

vsergiu93 replied:
Yep, but we need to keep this limitation in mind, if that's ok, I can try to use it and see how it goes.

brendt replied:
I think that's ok :)

brendt commented:
Awesome!
